### PR TITLE
Upgrading to CMake 3.0

### PR DESCRIPTION
**Pull request by Naios**
# Upgrading to CMake 3.0
- Uses inherited dependencies and awesome new features now as described in [this guide](http://rix0r.nl/blog/2015/08/13/cmake-guide/).
- Removes some workarrounds which targeted lower CMake versions
- The cleaner dependency tree improves the portability between 6.x <-> 3.3.5, all project use automatic inclusion now which makes it easier to add new sources.
  - New macros were added to auto include subdirectories and sources with an ignore list (for precompiled headers and platform dependent sources).
- Platform specific library linking handling was moved into the `dep` directory, libraries which aren't available on some platforms provide a dummy target which moves most of the crossplatform platform dependent logic into `dep` -> it's also easier to exchange libraries now.

**Target branch(es)**: 335/6x

**Issues addressed**:
- Lots of dependency clutter in the whole cmake buildfiles -> no proper usage of libraries as dependencies.
- Difficult "possible future linking to shared libraries" -> this patch makes it much easier through the usage of cmakes internal `BUILD_SHARED_LIBS` option.
- Circular dependency between game <-> scripts, the dependency graph is direct acyclic now.

**Tests performed**:
- Builds with compiler listed below

**Known issues and TODO list**:

**Test builds:**
- [x] Builds with MSVC 2013
- [x] Builds with MSVC 2015
  - [x] Incremental linking warnings
  - [x] No PCH build
- [x] Builds with Ubuntu GCC (with jemalloc)
- [x] Builds with Ubuntu Clang (with jemalloc)
- [x] Builds with MacOSX Clang  (with jemalloc)
  - [x] AppleClang detection

### CODE REVIEW

**Comment by Naios** [comment]
GCC with PCH:

```
cc1plus: warning: src/server/game/gamePCH.h.gch: not used because `BOOST_CHRONO_NO_LIB' is defined [-Winvalid-pch]
cc1plus: error: one or more PCH files were found, but they were invalid
compilation terminated due to -Wfatal-errors.
cc1plus: warning: src/server/game/gamePCH.h.gch: not used because `BOOST_CHRONO_NO_LIB' is defined [-Winvalid-pch]
cc1plus: error: one or more PCH files were found, but they were invalid
compilation terminated due to -Wfatal-errors.
src/server/game/CMakeFiles/game.dir/build.make:100: recipe for target 'src/server/game/CMakeFiles/game.dir/Handlers/AuthHandler.cpp.o' failed
make[2]: *** [src/server/game/CMakeFiles/game.dir/Handlers/AuthHandler.cpp.o] Error 1
cc1plus: warning: src/server/game/gamePCH.h.gch: not used because `BOOST_CHRONO_NO_LIB' is defined [-Winvalid-pch]
make[2]: *** Waiting for unfinished jobs....
cc1plus: error: one or more PCH files were found, but they were invalid
compilation terminated due to -Wfatal-errors.
src/server/game/CMakeFiles/game.dir/build.make:146: recipe for target 'src/server/game/CMakeFiles/game.dir/Handlers/PetHandler.cpp.o' failed
```

**Comment by Naios** [comment]
The GCC build with PCH was fixed, it would be nice if anyone could test this out (through building this PR with different configurations).
If the build succeeds everything is fine.

File: cmake/compiler/msvc/settings.cmake [hunk 1]
```diff
@@ -84,3 +84,16 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Zm500")
 # 'function' : member function does not override any base class virtual member function
 # 'virtual_function' : no override available for virtual member function from base 'class'; function is hidden
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /we4263 /we4264")
+
+# Disable incremental linking in debug builds
```
Comment (DDuarte) [inline cmake/compiler/msvc/settings.cmake]: Perhaps worth adding to the comment that this is needed due to linking getting stuck on VS (which might be fixed in a later version).

Comment (Naios) [inline cmake/compiler/msvc/settings.cmake]: Yep thanks, didn't know what the issue was exactly... but now I will add it.

**Comment by Gacko** [comment]
I'm currently testing Non PCH on OS X. Output inc.

**Comment by Gacko** [comment]
clang (installed)
```
$ clang --version
Apple LLVM version 7.0.2 (clang-700.1.81)
Target: x86_64-apple-darwin15.3.0
Thread model: posix
```
Make (installed)
```
$ make --version
GNU Make 3.81
Copyright (C) 2006  Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

This program built for i386-apple-darwin11.3.0
```
CMake (self-built)
```
$ cmake --version
cmake version 3.4.3

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```
OpenSSL (self-built)
```
$ openssl
OpenSSL> version
OpenSSL 1.0.2g  1 Mar 2016
```
CMake parameters
```
cmake $TRINITY/repo
-DREADLINE_INCLUDE_DIR=$TRINITY/include
-DREADLINE_LIBRARY=$TRINITY/lib/libreadline.dylib
-DBOOST_INCLUDEDIR=$TRINITY/include
-DBOOST_LIBRARYDIR=$TRINITY/lib
-DOPENSSL_SSL_LIBRARIES=$TRINITY/lib/libssl.dylib
-DOPENSSL_CRYPTO_LIBRARIES=$TRINITY/lib/libcrypto.dylib
-DOPENSSL_INCLUDE_DIR=$TRINITY/include
-DCMAKE_INSTALL_PREFIX=$TRINITY
-DWITH_WARNINGS=1
-DTOOLS=1
-DUSE_COREPCH=0
-DUSE_SCRIPTPCH=0
```
CMake output
```
-- The C compiler identification is AppleClang 7.0.2.7000181
-- The CXX compiler identification is AppleClang 7.0.2.7000181
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Check for working C compiler: /Library/Developer/CommandLineTools/usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++
-- Check for working CXX compiler: /Library/Developer/CommandLineTools/usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Detected 64-bit platform
-- UNIX: Using default configuration directory
-- UNIX: Using default library directory
-- UNIX: Configuring uninstall target
-- UNIX: Created uninstall target
-- UNIX: Detected compiler: /Library/Developer/CommandLineTools/usr/bin/cc
-- Using mysql-config: /usr/local/mysql/bin/mysql_config
-- Found MySQL library: /usr/local/mysql/lib/libmysqlclient.dylib
-- Found MySQL headers: /usr/local/mysql/include
-- Found MySQL executable: /usr/local/mysql/bin/mysql
-- Found git binary : /usr/bin/git

* TrinityCore revision   : 6858c43e7eb2 2016-03-17 17:07:42 +0100 (cmake3 branch)
* TrinityCore buildtype  : Release

* Install core to        : /Users/marco/Trinity
* Install libraries to   : /Users/marco/Trinity/lib
* Install configs to     : /Users/marco/Trinity/etc

* Build world/auth       : Yes (default)
* Build with scripts     : Yes (default)
* Build map/vmap tools   : Yes
* Build core w/PCH       : No
* Build scripts w/PCH    : No
* Show all warnings      : Yes
* Use coreside debug     : No  (default)
* Show source tree       : Yes - hierarchical
* Use GIT revision hash  : Yes

-- Looking for pthread.h
-- Looking for pthread.h - found
-- Looking for pthread_create
-- Looking for pthread_create - found
-- Found Threads: TRUE  
-- Boost version: 1.60.0
-- Found the following Boost libraries:
--   system
--   filesystem
--   thread
--   program_options
--   iostreams
--   regex
-- Performing Test boost_filesystem_copy_links_without_NO_SCOPED_ENUM
-- Performing Test boost_filesystem_copy_links_without_NO_SCOPED_ENUM - Success
-- Found ZLIB: /usr/lib/libz.dylib (found version "1.2.5") 
-- Looking for open
-- Looking for open - found
-- Found OpenSSL library: /Users/marco/Trinity/lib/libssl.dylib;/Users/marco/Trinity/lib/libcrypto.dylib
-- Found OpenSSL headers: /Users/marco/Trinity/include
-- Found Readline library: /Users/marco/Trinity/lib/libreadline.dylib
-- Include dir is: /Users/marco/Trinity/include
-- Found BZip2: /usr/lib/libbz2.dylib (found version "1.0.6") 
-- Looking for BZ2_bzCompressInit
-- Looking for BZ2_bzCompressInit - found
-- SCRIPT PREPARATIONS
--   -> Prepared: Spells
--   -> Prepared: Commands
--   -> Prepared: Custom
--   -> Prepared: World
--   -> Prepared: OutdoorPvP
--   -> Prepared: EasternKingdoms
--   -> Prepared: Kalimdor
--   -> Prepared: Outland
--   -> Prepared: Northrend
--   -> Prepared: Events
--   -> Prepared: Pet
-- SCRIPT PREPARATION COMPLETE

-- Configuring done
-- Generating done
-- Build files have been written to: /Users/marco/Trinity/build
```
Make log: http://pastebin.com/raw/tVQDXsri

**Comment by Gacko** [comment]
Non PCH on OS X works fine now.

**Comment by Rochet2** [comment]
I tested this on https://github.com/TrinityCore/TrinityCore/pull/16795/commits/d6b42f4d4a8af43cfcd9b93416757ba95486f28e and there were these warnings for each project. The build was VS2015+PCH+Debug.
```
Warning	LNK4044	unrecognized option '/INCREMENTAL:NO'; ignored	game	E:\Cores\Fork\BUILD\src\server\game\LINK	1	
Warning	LNK4044	unrecognized option '/INCREMENTAL:NO'; ignored	scripts	E:\Cores\Fork\BUILD\src\server\scripts\LINK	1
```
Here are the full outputs: https://gist.github.com/Rochet2/7d7c875ee7bf724d92e9

Here is another build that had errors on the same version. Build was VS2015+no PCH+Release.
https://gist.github.com/Rochet2/3bf34e10b5c66248824a
Here are the errors:
```
Severity	Code	Description	Project	File	Line	Suppression State
Error	C2614	'AsyncAcceptor': illegal member initialization: '_closed' is not a base or member (compiling source file E:\Cores\Fork\source\src\server\authserver\Main.cpp)	authserver	e:\cores\fork\source\src\server\shared\networking\AsyncAcceptor.h	35	
Error	C2039	'atomic': is not a member of 'std' (compiling source file E:\Cores\Fork\source\src\server\authserver\Main.cpp)	authserver	e:\cores\fork\source\src\server\shared\networking\AsyncAcceptor.h	84	
Error	C2143	syntax error: missing ';' before '<' (compiling source file E:\Cores\Fork\source\src\server\authserver\Main.cpp)	authserver	e:\cores\fork\source\src\server\shared\networking\AsyncAcceptor.h	84	
Error	C4430	missing type specifier - int assumed. Note: C++ does not support default-int (compiling source file E:\Cores\Fork\source\src\server\authserver\Main.cpp)	authserver	e:\cores\fork\source\src\server\shared\networking\AsyncAcceptor.h	84	
Error	C2238	unexpected token(s) preceding ';' (compiling source file E:\Cores\Fork\source\src\server\authserver\Main.cpp)	authserver	e:\cores\fork\source\src\server\shared\networking\AsyncAcceptor.h	84	
Error	C2065	'_closed': undeclared identifier (compiling source file E:\Cores\Fork\source\src\server\authserver\Main.cpp)	authserver	e:\cores\fork\source\src\server\shared\networking\AsyncAcceptor.h	70	
Error	C2228	left of '.exchange' must have class/struct/union (compiling source file E:\Cores\Fork\source\src\server\authserver\Main.cpp)	authserver	e:\cores\fork\source\src\server\shared\networking\AsyncAcceptor.h	70	
Error	C3861	'CancelSynchronousIo': identifier not found	worldserver	E:\Cores\Fork\source\src\server\worldserver\Main.cpp	303	
```

**Comment by Naios** [comment]
@Rochet2 Thanks, I could reproduce and fix the error.
I will also take a look at the incremental warning.

**Pull request merged**